### ISSUE TITLE
start-devicetree-overlay.sh

### ISSUE THREAD

**Issue by dementiev-net** [hzeller/ldgraphy#5]
**BBB:**

debian@beaglebone:~/ldgraphy/device-tree$ sudo ./start-devicetree-overlay.sh LDGraphy.dts
make: 'LDGraphy-00A0.dtbo' is up to date.

Adding LDGraphy overlay
./start-devicetree-overlay.sh: line 69: /sys/devices/bone_capemgr.*/slots: No such file or directory
cat: '/sys/devices/bone_capemgr.*/slots': No such file or directory
debian@beaglebone:~/ldgraphy/device-tree$

**Comment by VasUtkin** [comment]
another problem:

make: 'LDGraphy-00A0.dtbo' is up to date.
Adding LDGraphy overlay
./start-devicetree-overlay.sh: line 69: echo: write error: Invalid argument
 0: PF----  -1 
 1: PF----  -1 
 2: PF----  -1 
 3: PF----  -1 
debian@beaglebone:~/ldgraphy/device-tree$

**Comment by hzeller** [comment]
are you running a 3.18 kernel?

**Comment by dementiev-net** [comment]
Latest version: 4.9.82-ti-r102
(Debian 9.3 2018-03-05 4GB SD IoT)

**Comment by hzeller** [comment]
yeah, they changed the way overlays work on newer kernels. Either you adapt that or switch to a distribution with an older kernel.

**Comment by atmelletje** [comment]
> yeah, they changed the way overlays work on newer kernels. Either you adapt that or switch to a distribution with an older kernel.

Hi Hzeller,
I looked at this for a long time and now is the time (as I now have more free time) to start with it. I'm a mechanical engineer but do not have much clue about software repo's distros.  could you tell me what version distri will work as there are many of them.
my BBB is a rev C with 4+ kernel debian flashed.  thank you in advance

g++ -std=c++0x -Wall -W -Wno-unused-function -I../am335x_pru_package/pru_sw/app_loader/include -D_XOPEN_SOURCE=500 -mtune=cortex-a8 -march=armv7-a -O3 -c  image-processing.cc -o image-processing.o
image-processing.cc:21:17: fatal error: png.h: No such file or directory
 #include <png.h>
                 ^
compilation terminated.
Makefile:50: recipe for target 'image-processing.o' failed
make: *** [image-processing.o] Error 1
debian@beaglebone:/var/lib/cloud9/ldgraphy/src$


2nd.

Package libpng-dev is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'libpng-dev' has no installation candidate
debian@beaglebone:/var/lib/cloud9/ldgraphy/src$


Roel

**Comment by hstarmans** [comment]
Hey @atmelletje,

I worked on ldgraphy for some time and then moved to Amaranth HDL.
You could check out the library here (https://github.com/hstarmans/ldgraphy/tree/master).

Something you could look into are;
   1) how do i blink blink the laser using a PRU
        https://github.com/hstarmans/ldgraphy/tree/master/scripts/pru_tests/blinklaser_channel1
```python
#!/usr/bin/python3
""" blinklaser1.py - test script for the Firestarter
blinks the laser 3 times with a period of 6 seconds
"""python
from uio.ti.icss import Icss

pruss = Icss('/dev/uio/pruss/module')
pruss.initialize()
core = pruss.core0
core.load('blinklaser1.bin')
core.run()
print('Waiting for core to halt')
while not core.halted:
    pass
```
After the laser test I would run;
  - test polygon https://github.com/hstarmans/ldgraphy/tree/master/scripts/pru_tests/spinpolygon
  - test diode https://github.com/hstarmans/ldgraphy/tree/master/scripts/pru_tests/photodiodetest
  - try to stabilize https://github.com/hstarmans/ldgraphy/tree/master/scripts/pru_tests/stabilizer

Note that in my install notes, 
  i tell you exactly which image I used 
https://rcn-ee.net/rootfs/bb.org/testing/2019-09-08/stretch-iot/bone-debian-9.10-iot-armhf-2019-09-08-4gb.img.xz

This could be a good way to start, good luck!

**Comment by atmelletje** [comment]
Thanks Hstarmans. dit helpt me op de juiste weg te komen echter de img is niet voorhanden daar !

Roel

**Comment by hstarmans** [comment]
Your welcome!  Henner zeller wilde uiteindelijk ook naar een FPGA gaan en heeft hier nog een tijdje aan gewerkt. Dit werk staat in de beagleg repository in de FPGA branch. 
Merk op dat de PRU route nadelen heeft, ik zou hier wel rekening mee houden.

Ik heb er behoorlijk wat tijd ingestorken en mijn oplossing staat nu hier https://github.com/hstarmans/hexastorm/blob/master/src/hexastorm/lasers.py
De python code wordt geconverteerd naar HDL en dit gaat dan de FPGA in via icestorm.
De variant waar ik momenteel aan werk heeft de FPGA op de printplaat en de motor is verwerkt in het PCB.

![screenshot](https://github.com/hzeller/ldgraphy/assets/6517246/37d5f63c-3e1e-4a2e-8f39-78cf76efdbff)

De lenzen zijn optioneel maar geven wel een significante prestatie verbetering.
Ik verwacht een meer definitieve versie over een maand te hebben, ik  ga nog op vakantie.

laat het weten als je vragen hebt, ik gebruik mijn beagle en de bijbehorende printplaten niet meer.

**Comment by atmelletje** [comment]
Hey Rik,

ik ben geen code writer, moet et meer van een werkende installatie hebben. ik ga het op een andere manier gebruiken dmv een drum met film te beichten.  (filmbelichter/photplotter) mijn intentie is om het voor meerdere toepassingen te kunnen gebruiken zoals Pad Printing. waarbij de "hexagon/square" de drum zal zijn. Dat is de bedoeling hiervan.

Anyways! installation fails because of a missing file ( libpng-dev)   

g++ -std=c++0x -Wall -W -Wno-unused-function -I../am335x_pru_package/pru_sw/app_loader/include -D_XOPEN_SOURCE=500 -mtune=cortex-a8 -march=armv7-a -O3 -c  image-processing.cc -o image-processing.o
image-processing.cc:21:17: fatal error: png.h: No such file or directory
 #include <png.h>
                 ^
compilation terminated.
Makefile:50: recipe for target 'image-processing.o' failed
make: *** [image-processing.o] Error 1


Rik
"laat het weten als je vragen hebt, ik gebruik mijn beagle en de bijbehorende printplaten niet meer."
ben niet in Nederlad anders had ik et graag willen gebruiken, is het mogelijk dat je een img kan maken van de werkende distri?

Roel

**Comment by hstarmans** [comment]
@atmelletje, did you install the relevant packages as outlined by Henner Zeller?

**From the documentation;**
Install relevant packages. We are reading png-images as input, so we need the library that helps us with that (probably already installed on your system)
```
sudo apt-get update
sudo apt-get install libpng-dev -y
```
There are also other solutions as turning part of the code into comments, changing the compilation order etc. This might bring you closer to identifying the issue.

Mechanical skills can be very useful for this project but you will also need extensive electronic and software skills.  If you have a different setup, you will for sure need to adapt the code.
There is company in Eindhoven which used to illuminate a rotating drum with a laser (4Pico), although i can no longer find that system on their site.

I can imagine that with a high motivation and a willingness to learn to code, you can end up with some sort of setup in a couple of months / weeks.
Still, it will be hard to scale or monetize and most likely is doomed.

**Comment by atmelletje** [comment]
the problem is that the Libpng-dev can't be found with an update. during the update, several source packages return an error. 
I know enough about electronics and mechanics to get this to work.  

Croot@beaglebone:/var/lib/cloud9#sudo apt-get install libpng-dev -y
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package libpng-dev
root@beaglebone:/var/lib/cloud9#

I'm Dutch, I only give up when I bite the dust. :)

**Comment by hstarmans** [comment]
There are roughly two options;
   1) find a compiled binary, look online for your distro and platform and add a source.
       in ubuntu this is known as adding a ppa.
   2) create a binary yourself by compiling libpng from its source code, this possible takes longer and is a 
       bit more complicated.

I am sure you have the skills and determination to do it, but given the fact you did not build one makes you underestimate the time required.
The devil is also in the last 20 percent. Once you reach proof of concept stage you will realize you still need to do 80 percent of the work.

**Comment by atmelletje** [comment]
> i tell you exactly which image I used
https://rcn-ee.net/rootfs/bb.org/testing/2019-09-08/stretch-iot/bone-debian-9.10-iot-armhf-2019-09-08-4gb.img.xz

this image has vanished :)

**Comment by hstarmans** [comment]
@atmelletje, I would go to a recent image. My code mainly relies on the work of mvduin (another dutchmen :-))
https://github.com/mvduin/py-uio
He actually takes care of the interaction between scrips that are run on top of the beagle bone PRU.
His work is also maintained. You should then be able to test the hardware.
You can also sent me an email and I can pick it up from there; hstarmans@hexastorm.com.

I am in Mexico at the moment, so there can be some delay.
